### ISSUE TITLE
Device widget icons for block attached device inexplicably showing plus instead of eject icon (current-testing).

### ISSUE THREAD

**Issue by brendanhoar** [QubesOS/qubes-issues#4849]
### Qubes OS version:

R4.0
( w/ current testing repository dom0 and all templates up to date as of 2/26/2019 )

### Affected component(s) or functionality:

Device attach widget. In particular a partition block devices' attachment status indicator.

---

### Steps to reproduce the behavior:

1. Attach device to a VM by selecting partition in widget and navigating to submenu item with a plus and the vm name you wish to attach it to, which you click. 
2. Mount and use partition in that VM. 
3. When finished, unmount partition. 
4. Before shutting down VM, attempt to use widget to detach device by navigating to the partition device, locating the VM, noting the indicator of attachment and selecting the VM to cause Qubes to detach the device from the VM.

### Expected or desired behavior:

Drop down menu should show the VM it is attached to with the special ("eject?") icon indicating the device is attached and you can detach it. This has been the case for quite a while.

### Actual behavior:

Drop down menu surprisingly shows a *plus sign* next to the VM name even though that partition device is attached to the VM. Due to being unsure what would happen if I selected this (is this a display error? Or does Xen think the device is not attached?), I did not select the item. Instead I shut down the VM and did not see any errors. System seemed in a normal state.

### General notes:

This happened twice in the past couple of days. As far back as I can remember, I have not seen this issue before.

While this very well could be a display bug, out of an abundance of caution I decided not to explore too much farther with my primary Qubes system.

---

### I have consulted the following relevant [documentation](https://www.qubes-os.org/doc/):

The only changes recently pushed that I can see that appear to be likely to relate are these changes to deal with device list display issues when the widgets are at the bottom of the screen. The links are the original and the merge. It could just be coincidence...but...
https://github.com/QubesOS/qubes-desktop-linux-manager/commit/11cf897
https://github.com/QubesOS/qubes-desktop-linux-manager/commit/3a2d229

**Comment by brendanhoar** [comment]
After a Qubes session yesterday, I can confirm this is a bug in tracking the state of the attachment (vs "just" a display issue).

Before shutting down the VM, out of habit I navigated to device-partition menu item and vmname sub-menu item and let go of the button...while at the same time noticing that the vm had a plus sign instead of the eject indicator.

I received an notification that stated the device was already attached to the VM.

Brendan

**Comment by mvermaes** [comment]
Just saw this as well, after rebooting host for latest dom0 and Fedora 28 updates, and attaching/attempting to detach a USB block device for the first time. Not using any testing repos.

**Comment by mvermaes** [comment]
Killing `qui-devices` per https://github.com/QubesOS/qubes-issues/issues/4878#issuecomment-472175527 and letting it restart causes the device to appear connected again, allowing it to be detached.

**Comment by yubiguel** [comment]
> Killing `qui-devices` per [#4878 (comment)](https://github.com/QubesOS/qubes-issues/issues/4878#issuecomment-472175527) and letting it restart causes the device to appear connected again, allowing it to be detached.

However, even after killing qui-devices problems on attaching and detaching further usb devices persist.

**Comment by slq32** [comment]
Same here, killing qui-devices and letting it restart temporarily fixes the problem (until a new device is connected or attached/detached. When trying to attach i sometimes get "Error: Got empty response from qubesd. See journalctl in dom0 for details"

In dom0 i get this, maybe it's helpful: 

```Mar 17 12:56:02 dom0 qrexec[7566]: qubes.USB+2-3: print-scan -> sys-usb: allowed to sys-usb
Mar 17 12:56:02 dom0 qubesd[1793]: unhandled exception while calling src=b'dom0' meth=b'admin.vm.device.usb.Attach' dest=b'print-scan' arg=b'sys-usb+2-3' len(untrusted_payload)=0
Mar 17 12:56:02 dom0 qubesd[1793]: Traceback (most recent call last):
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib/python3.5/site-packages/qubes/api/__init__.py", line 266, in respond
Mar 17 12:56:02 dom0 qubesd[1793]:     untrusted_payload=untrusted_payload)
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib64/python3.5/asyncio/futures.py", line 381, in __iter__
Mar 17 12:56:02 dom0 qubesd[1793]:     yield self  # This tells Task to wait for completion.
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib64/python3.5/asyncio/tasks.py", line 310, in _wakeup
Mar 17 12:56:02 dom0 qubesd[1793]:     future.result()
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib64/python3.5/asyncio/futures.py", line 294, in result
Mar 17 12:56:02 dom0 qubesd[1793]:     raise self._exception
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib64/python3.5/asyncio/tasks.py", line 240, in _step
Mar 17 12:56:02 dom0 qubesd[1793]:     result = coro.send(None)
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib/python3.5/site-packages/qubes/api/admin.py", line 1248, in vm_device_attach
Mar 17 12:56:02 dom0 qubesd[1793]:     yield from self.dest.devices[devclass].attach(assignment)
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib/python3.5/site-packages/qubes/devices.py", line 254, in attach
Mar 17 12:56:02 dom0 qubesd[1793]:     device=device, options=device_assignment.options)
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib/python3.5/site-packages/qubes/events.py", line 236, in fire_event_async
Mar 17 12:56:02 dom0 qubesd[1793]:     effect = task.result()
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib64/python3.5/asyncio/futures.py", line 294, in result
Mar 17 12:56:02 dom0 qubesd[1793]:     raise self._exception
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib64/python3.5/asyncio/tasks.py", line 240, in _step
Mar 17 12:56:02 dom0 qubesd[1793]:     result = coro.send(None)
Mar 17 12:56:02 dom0 qubesd[1793]:   File "/usr/lib/python3.5/site-packages/qubesusbproxy/core3ext.py", line 259, in on_device_attach_usb
Mar 17 12:56:02 dom0 qubesd[1793]:     device.backend_domain.name))
Mar 17 12:56:02 dom0 qubesd[1793]: ValueError: list.remove(x): x not in list
```

However the device does get attached/detached...

**Comment by Eric678** [comment]
Thank you Marta, been hanging out for this update for ages.  Great to have a working devices widget again - all looks OK that I could test, including working with the new persistent USB updates in dom0. I have installed it on my -current system using (for anyone else who cannot wait):
`qubes-dom0-update --action='-v update' --enablerepo=qubes-dom0-current-testing qubes-desktop-linux-manager`
Odd had to use that action else it was skipped due to missing dependencies?

Andrew: writing here because my posts to the list are getting quarantined - you need to add to your job description keeping a "Known Issues" page near the top of docs up to date in plain English (with workarounds) the devices widget is a prime example, the other one that deserves to be there is the X focus issue: 1 VM across multiple desktops. Github is not very useful for ordinary users, was ages before I stumbled upon the open issue. It is quite disconcerting having mouse over popups and right click menus appear from a desktop I just left.

thanks, Eric

**Comment by andrewdavidwong** [comment]
@Eric678: 
> Andrew: writing here because my posts to the list are getting quarantined

I've just approved your message (from ~6 hours ago) in the qubes-users moderation queue. In the future, please be patient while we get to the moderation queue. We usually check it at least once a day.

> you need to add to your job description keeping a "Known Issues" page near the top of docs up to date in plain English (with workarounds) the devices widget is a prime example, the other one that deserves to be there is the X focus issue: 1 VM across multiple desktops. Github is not very useful for ordinary users, was ages before I stumbled upon the open issue. It is quite disconcerting having mouse over popups and right click menus appear from a desktop I just left.

I appreciate the feedback, but this sort of thing never works out as well as it sounds. Maintaining a constantly updated list of issues in text and ensuring that it's actually useful inevitably results in trying to re-create an issue tracker by hand. This is exactly why dedicated software issue trackers exist and why we use one. (Not to mention that which issues count as important enough to appear on the document vary from user to user.)

It would be much better for us to make qubes-issues user-friendly enough to be the one place where issues are created, discovered, and updated. There are three main ways we can do this:

1. Issue creators and moderators can ensure that issue titles and descriptions are clearly written with search in mind.
2. Issue creators, devs, and moderators can ensure that issues are updated when things change.
3. Users can search the issue tracker when they have questions and experience problems.

If you have further suggestions about how we can improve qubes-issues, please feel free to open an issue for it or submit a change to [Reporting bugs and other issues](https://www.qubes-os.org/doc/reporting-bugs/).

**Comment by brendanhoar** [comment]
Not sure this is entirely addressed. Updated to current-testing this afternoon to fix the VM list restarting in an infinite loop. Attached a device to a VM, yay, the plus sign appeared. Shut down the VM...device menu item still bolded and still showed attached to that VM, even though it wasn't running. Killed off the qui-devices, systemd restarted it, started a VM and tried to attach the device to that VM...but now that VM wasn't in the drop down list. Killed and restarted qui-devices, and now I could attach it to the VM. 

:/

B

**Comment by yubiguel** [comment]
Everything seems to be working much better. The only remaining issue is when you shutdown a vm that has an attached device and then restart that vm with the device still attached. The device manager will continue to say the device is attached but it is not. and if you try to attach it gives an error. It is not a big headache as the simple solution is to physically remove the device and then re-connect it, this fixes the issue as once the device is physically removed then the device manager correctly shows it detached.

**Comment by brendanhoar** [comment]
How does the widget keep in sync with command-line attach/detach requests and results? Can the widget be notified when a VM exits? Can the widget poll for the start time of VMs?

Any/all via dbus? Anything not covered that might require polling?

B

**Comment by marmarta** [comment]
@brendanhoar the biggest problem is that some events are not fired by design (when a VM is starting or halting, devices are not reported as being attached/detached); I've pushed a fix for that, but hopefully one day the whole device API will get a needed rewrite (won't it, @marmarek  ?)

**Comment by yubiguel** [comment]
I'm experiencing a strange issue during testing, at times attaching a device (yubikey) does nothing at all. The device does not get attached. Although the device does show in the list it simply does not get attached to the vm. The only solution is to kill qui-devices which then restarts itself and then its possible to attach again, or to attach the device via command line (qvm-usb)

**Comment by marmarek** [comment]
I've created new issue (#4999) about this, since this one is about the case when attaching works but is displayed incorrectly.

**Comment by slq32** [comment]
I can confirm this is now fixed, at least for me.
However i have another issue which i'm not sure it's related to this one: Neither dom0 nor sys-usb "sees" any device connected to the usb-c port. Thunderbolt is disabled in the bios and i guess it should be seen as a regular usb3 port. I get this:
```
[Tue Apr 30 09:54:32 2019] ACPI Error: [SPRT] Namespace lookup failure, AE_ALREADY_EXISTS (20170728/dswload2-346)
[Tue Apr 30 09:54:32 2019] ACPI Exception: AE_ALREADY_EXISTS, During name lookup/catalog (20170728/psobject-252)
[Tue Apr 30 09:54:32 2019] ACPI Error: Method parse/execution failed \_GPE.XTBT, AE_ALREADY_EXISTS (20170728/psparse-550)
[Tue Apr 30 09:54:32 2019] ACPI Error: Method parse/execution failed \_GPE.XTBT, AE_ALREADY_EXISTS (20170728/psparse-550)
[Tue Apr 30 09:54:32 2019] ACPI Error: Method parse/execution failed \_GPE._E42, AE_ALREADY_EXISTS (20170728/psparse-550)
[Tue Apr 30 09:54:32 2019] ACPI Exception: AE_ALREADY_EXISTS, while evaluating GPE method [_E42] (20170728/evgpe-646)
```
in dom0 and nothing in sys-usb. Booting Ubuntu live cd i get the same ACPI errors, but the device is recognized ... No idea if it's a Qubes issue or a Xen issue or if someone else complained before.
lspci shows a signe USB controller:
```
00:14.0 USB controller: Intel Corporation Sunrise Point-LP USB 3.0 xHCI Controller (rev 21)
```
which is attached to sys-usb.

**Comment by marmarek** [comment]
I'm not sure about your case, but I've seen on some machines (Thinpad) that USB-C related USB controller dynamically appear only when you plug some device in. And since we've disabled PCI hotplug, you won't see it.

**Comment by slq32** [comment]
This is a Dell laptop, but i'm guessing it's the same thing. It's not a big deal anyway.
Thanks !

**Comment by brendanhoar** [comment]
Update, using -testing:

I am still occasionally finding that after I start a VM, wait for a terminal window to appear, and select device assignment of a partition through the drop down, the assignment silently fails and the status in the drop down remains in sync with the current assignment (nothing is shown to be assigned). Repeated attempts continue to silently fail to assign the partition device.

I then restart the target VM, wait for the terminal window to appear and then retry assigning through the drop down, and...it works and is properly displayed in the drop down.

**Comment by marmarta** [comment]
I have trouble reproducing the error, so I've pushed a small fix to make the widget report more errors (after all, a fail is a fail and it should not be silent) - once it is merged, hopefully we'll see what exactly is happening.
